### PR TITLE
Fix for #3056

### LINKED ISSUE #3056: Destination array was not long enough. Check the destination index, length, and the array's lower bounds
```
Destination array was not long enough. Check the destination index, length, and the array's lower bounds. 

   at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)\r\n   at LinqToDB.Metadata.MetadataReader.GetDynamicColumns(Type type)\r\n   at LinqToDB.Mapping.MappingSchema.GetDynamicColumns(Type type)\r\n   at LinqToDB.Mapping.EntityDescriptor.Init()\r\n   at LinqToDB.Mapping.EntityDescriptor..ctor(MappingSchema mappingSchema, Type type)\r\n   at LinqToDB.Mapping.MappingSchema.<>c.<GetEntityDescriptor>b__94_0(ICacheEntry`1 o, MappingSchema context)\r\n   at LinqToDB.Common.Internal.Cache.CacheExtensions.GetOrCreate[TItem,TKey,TContext](IMemoryCache`1 cache, TKey key, TContext context, Func`3 factory)\r\n   at LinqToDB.Mapping.MappingSchema.GetEntityDescriptor(Type type)\r\n   at LinqToDB.Linq.Table`1.InitTable(IDataContext dataContext, Expression expression)\r\n   at LinqToDB.Linq.Table`1..ctor(IDataContext dataContext)\r\n   at LinqToDB.Data.DataConnection.GetTable[T]()\r\n   at LinqToDB.Data.DataConnectionExtensions.BulkCopy[T](DataConnection dataConnection, BulkCopyOptions options, 
```

### Steps to reproduce
in version 3.3 - it's working
in version 3.4 - error

```c#
    [TestClass]
    public class DataModelTest
    {
       [TestMethod]
        public void DataModelDynamicTableTest2()
        {
            using (var dm = new DataModel("SqlServer", "Data Source=.;Initial Catalog=testdb;Integrated Security=sspi;"))
            {
                dm.CommandTimeout = 50;
                dm.DropTable<TestRow>(throwExceptionIfNotExists: false);
                dm.CreateTable<TestRow>();

                var fm = MappingSchema.Default.GetFluentMappingBuilder();

                fm.Entity<DynamicTableRow>()
                    //.HasIdentity(x => Sql.Property<int>(x, "Id"))
                    .Property(x => Sql.Property<string>(x, "Name"))
                    .Property(x => Sql.Property<string>(x, "Description"));

                var rows = new List<DynamicTableRow>();
                var drow = new DynamicTableRow();
                drow.Properties.Add("Id", 1);
                drow.Properties.Add("Name", "n1");
                drow.Properties.Add("Description", "d0");
                rows.Add(drow);
                drow = new DynamicTableRow();
                drow.Properties.Add("Id", 2);
                drow.Properties.Add("Name", "n2");
                drow.Properties.Add("Description", "d00");
                rows.Add(drow);

                dm.BulkCopy(new BulkCopyOptions
                {
                    TableName = "TestRow",
                    SchemaName = "dbo",
                    MaxBatchSize = 100
                },
                 rows);
            }
        }
    }

    [Table("TestRow", Schema = "dbo")]
    class TestRow
    {
        [PrimaryKey, Identity]
        public int Id;

        [Column(DataType = DataType.VarChar, Length = 100)]
        public string Name;

        [Column(DataType = DataType.VarChar, Length = 500)]
        public string Description;

    }

    [Table]
    public class DynamicTableRow
    {
        [DynamicColumnsStore]
        public Dictionary<string, object> Properties = new Dictionary<string, object>();
    }
```

### Environment details
linq2db version: 3.4.0
Database Server: SqlServer 2014
Database Provider: SqlServer
Operating system: Windows 10
.NET Framework: core 3.1

### PR DESCRIPTION
**Pull request by to11mtm**
Simple change, fix MetadataReader.GetDynamicColumns to not reset length variable until we have made array.

Fixes #3056

### CODE REVIEW

**Comment by sdanyliv** [comment]
@to11mtm, limit test to only SQLite.

**Comment by to11mtm** [comment]
@sdanyliv When I try to change to `TestProvName.AllSQLite`, The SQLite.MS tests wind up failing when I run all tests, but when I run just the MS tests it's fine;

```
System.ArgumentException : An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at LinqToDB.Mapping.EntityDescriptor.Init() in C:\github\linq2db\Source\LinqToDB\Mapping\EntityDescriptor.cs:line 231
   at LinqToDB.Mapping.EntityDescriptor..ctor(MappingSchema mappingSchema, Type type) in C:\github\linq2db\Source\LinqToDB\Mapping\EntityDescriptor.cs:line 31
   at LinqToDB.Mapping.MappingSchema.<>c.<GetEntityDescriptor>b__94_0(ICacheEntry`1 o, MappingSchema context) in C:\github\linq2db\Source\LinqToDB\Mapping\MappingSchema.cs:line 1593
   at LinqToDB.Common.Internal.Cache.CacheExtensions.GetOrCreate[TItem,TKey,TContext](IMemoryCache`1 cache, TKey key, TContext context, Func`3 factory) in C:\github\linq2db\Source\LinqToDB\Common\Internal\Cache\MemoryCacheExtensions.cs:line 120
   at LinqToDB.Mapping.MappingSchema.GetEntityDescriptor(Type type) in C:\github\linq2db\Source\LinqToDB\Mapping\MappingSchema.cs:line 1587
   at LinqToDB.Linq.Table`1.InitTable(IDataContext dataContext, Expression expression) in C:\github\linq2db\Source\LinqToDB\Linq\TableT.cs:line 32
   at LinqToDB.Linq.Table`1..ctor(IDataContext dataContext) in C:\github\linq2db\Source\LinqToDB\Linq\TableT.cs:line 20
   at LinqToDB.Data.DataConnection.GetTable[T]() in C:\github\linq2db\Source\LinqToDB\Data\DataConnection.Linq.cs:line 24
   at LinqToDB.Data.DataConnectionExtensions.BulkCopy[T](DataConnection dataConnection, BulkCopyOptions options, IEnumerable`1 source) in C:\github\linq2db\Source\LinqToDB\Data\DataConnectionExtensions.cs:line 2194
   at Tests.UserTests.Issue3056Tests.DataModelDynamicTableTest2(String context) in C:\github\linq2db\Tests\Linq\UserTests\Issue3056Tests.cs:line 38
```

I can set to just SQLiteMS or SQLiteClassic, but wasn't sure whether this was a separate concern we should look into?

File: Tests/Linq/UserTests/Issue3056Tests.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	[TestFixture]
+	public class Issue3056Tests : TestBase
+	{
+		[Test]
+		public void DataModelDynamicTableTest2([DataSources(false)] string context)
+		{
+			using (var db = new TestDataConnection(context))
+			using (db.CreateLocalTable<TestRow>())
+			{
+				var fm = MappingSchema.Default.GetFluentMappingBuilder();
```
Comment (sdanyliv) [inline Tests/Linq/UserTests/Issue3056Tests.cs:19]: Create new `MappingScgema` before DataConnection and pass this mapping schema to the constructor. `MappingSchema.Default` is singleton.

**Review by sdanyliv — APPROVED** [review]

File: Tests/Model/TestDataConnection.cs [hunk 2]
```diff
@@ -20,6 +21,12 @@ public TestDataConnection(string configString)
 //			if (configString == ProviderName.SqlServer2008 && ++_counter > 1000)
 //				OnClosing += TestDataConnection_OnClosing;
 		}
+		
+		public TestDataConnection(string configString, MappingSchema mappingSchema)
```
Comment (MaceWindu) [inline Tests/Model/TestDataConnection.cs:25]: This wasn't necessary, context getter method on TestBase already accepts mapping schema..

**Pull request merged**